File: mobile/lib/models/visitor_event.dart
import '../core/parse.dart';

/// One detected person at the door. Mirrors the backend `Person` dataclass.
/// KNOWN people are named only (no age/gender surfaced); UNKNOWN people carry a
/// cautious, approximate description.
class Person {
  final bool known;
  final String name;
  final double confidence;
  final int? age; // approximate; ONLY shown as a range, and never for known
  final String gender; // "man" | "woman" | ""
  final bool isSpoof;
  final double spoofScore; // 1.0 real -> 0.0 spoof
  final String appearance; // clothing / carried (unknown only)
  final String expression; // cautious mood cue (unknown only)

  const Person({
    required this.known,
    required this.name,
    required this.confidence,
    required this.age,
    required this.gender,
    required this.isSpoof,
    required this.spoofScore,
    required this.appearance,
    required this.expression,
  });

  factory Person.fromJson(Map<String, dynamic> j) => Person(
        known: asBool(j['known']),
        name: asStr(j['name'], 'Unknown'),
        confidence: asDouble(j['confidence']),
        age: asIntOrNull(j['age']),
        gender: asStr(j['gender']),
        isSpoof: asBool(j['is_spoof']),
        spoofScore: asDouble(j['spoof_score'], 1.0),
        appearance: asStr(j['appearance']),
        expression: asStr(j['expression']),
      );
}

/// A visitor event — the "spine" object. Tolerant of missing fields so older or
/// partial rows never break the UI.
class VisitorEvent {
  final String eventId;
  final String timestamp;
  final String trigger;

  // Primary identity (mirrors the largest/first-known person).
  final bool known;
  final String name;
  final double identityConfidence;

  final bool isSpoof;
  final double spoofScore;

  // Multi-person.
  final List<Person> people;
  final int extraUnknown;
  final int visitorCount;

  // Vision / scene.
  final List<String> carriedObjects;
  final String sceneSummary;
  final String hazards;
  final String ocrText;

  // Description (unknown-only, hedged).
  final int? age;
  final String gender;
  final String appearance;

  // Speech.
  final String speechTranscript;
  final String languageDetected;
  final String translatedTranscript;

  // Re-ID.
  final String? reidId;
  final int reidSeenCount;

  // Context.
  final String intent;
  final String announcementText;

  const VisitorEvent({
    required this.eventId,
    required this.timestamp,
    required this.trigger,
    required this.known,
    required this.name,
    required this.identityConfidence,
    required this.isSpoof,
    required this.spoofScore,
    required this.people,
    required this.extraUnknown,
    required this.visitorCount,
    required this.carriedObjects,
    required this.sceneSummary,
    required this.hazards,
    required this.ocrText,
    required this.age,
    required this.gender,
    required this.appearance,
    required this.speechTranscript,
    required this.languageDetected,
    required this.translatedTranscript,
    required this.reidId,
    required this.reidSeenCount,
    required this.intent,
    required this.announcementText,
  });

  factory VisitorEvent.fromJson(Map<String, dynamic> j) {
    final identity = asMap(j['identity']);
    final people = asMapList(j['people']).map(Person.fromJson).toList();
    return VisitorEvent(
      eventId: asStr(j['event_id']),
      timestamp: asStr(j['timestamp']),
      trigger: asStr(j['trigger'], 'manual'),
      known: asBool(identity['known']),
      name: asStr(identity['name'], 'Unknown'),
      identityConfidence: asDouble(identity['confidence']),
      isSpoof: asBool(j['is_spoof']),
      spoofScore: asDouble(j['spoof_score'], 1.0),
      people: people,
      extraUnknown: asInt(j['extra_unknown']),
      visitorCount: asInt(j['visitor_count']),
      carriedObjects: asStrList(j['carried_objects']),
      sceneSummary: asStr(j['scene_summary']),
      hazards: asStr(j['hazards'], 'none'),
      ocrText: asStr(j['ocr_text']),
      age: asIntOrNull(j['age']),
      gender: asStr(j['gender']),
      appearance: asStr(j['appearance']),
      speechTranscript: asStr(j['speech_transcript']),
      languageDetected: asStr(j['language_detected']),
      translatedTranscript: asStr(j['translated_transcript']),
      reidId: j['reid_id'] == null ? null : asStr(j['reid_id']),
      reidSeenCount: asInt(j['reid_seen_count']),
      intent: asStr(j['intent'], 'unknown visitor'),
      announcementText: asStr(j['announcement_text']),
    );
  }

  /// How many people the UI should present. Prefer visitor_count; fall back to
  /// people + extras so a partial event still reads sensibly.
  int get totalPeople {
    final byParts = people.length + extraUnknown;
    return visitorCount > 0 ? visitorCount : (byParts > 0 ? byParts : 0);
  }

  int get knownCount => people.where((p) => p.known && !p.isSpoof).length;

  int get unknownCount {
    final total = totalPeople;
    final k = knownCount;
    final u = total - k;
    return u < 0 ? 0 : u;
  }

  bool get anySpoof => isSpoof || people.any((p) => p.isSpoof);

  /// "3 people — 1 known, 2 unknown" (grammatical for 0/1/many).
  String get countLine {
    final total = totalPeople;
    if (total <= 0) return 'No one detected';
    final noun = total == 1 ? 'person' : 'people';
    final parts = <String>[];
    if (knownCount > 0) parts.add('$knownCount known');
    if (unknownCount > 0) parts.add('$unknownCount unknown');
    if (parts.isEmpty) return '$total $noun';
    return '$total $noun — ${parts.join(', ')}';
  }

  bool get hasSpeech =>
      speechTranscript.trim().isNotEmpty ||
      translatedTranscript.trim().isNotEmpty;
}

File: accessai/visitor_event.py
"""
The Visitor Event - the single data object that flows through the whole system.

This is the SPINE of AccessAI. Every AI module (present and future) WRITES into
this object; every output (voice, notification, history) READS from it.

Design rule for all 10 phases
-----------------------------
Adding a new AI capability = fill in a field that already exists here, and have
exactly one module responsible for it. You never restructure the pipeline to add
a feature. That is why this dataclass is defined in Phase 1 with fields reserved
for ALL later phases, even though Phase 1 leaves almost all of them at their
defaults.

Field -> phase that fills it:
    identity / face_box         -> Phase 2 (face recognition)
    visitor_count / objects     -> Phase 3 (object detection)
    carried_objects / intent    -> Phase 3 (context engine)
    announcement_text           -> Phase 4 (accessibility output)
    spoof_score / is_spoof      -> Phase 5 (anti-spoofing)
    scene_summary / ocr_text    -> Phase 6 (VLM + OCR)
    speech_transcript / language -> Phase 7 (speech recognition)
    translated_transcript       -> Phase 8 (translation)
    reid_id / reid_seen_count   -> Phase 9 (re-ID + auto-enroll)
    age / gender                -> Phase 12 (InsightFace genderage, approximate)
    appearance                  -> Phase 12 (VLM clothing/uniform, unknown only)
"""

from dataclasses import dataclass, field, asdict
from typing import List, Optional


@dataclass
class Identity:
    """Who the visitor is (filled by face recognition in Phase 2)."""
    known: bool = False
    name: str = "Unknown"
    confidence: float = 0.0


@dataclass
class Person:
    """ONE detected person at the door (Phase 15, multi-person).

    A doorbell frame can contain several people; the pipeline builds one Person
    per detected face and appends any face-less YOLO bodies as a count
    (extra_unknown on the event). KNOWN people are named only; UNKNOWN people
    carry a cautious, approximate description (age/gender from InsightFace,
    clothing/carried/expression from the VLM). `identity` on the event stays the
    PRIMARY person (first known, else largest) for full backward compatibility.
    """
    known: bool = False
    name: str = "Unknown"
    confidence: float = 0.0
    age: Optional[int] = None          # approximate InsightFace age, spoken as RANGE
    gender: str = ""                   # "man" | "woman" | "" (mapped from M/F)
    box: tuple = (0, 0, 0, 0)          # (x1, y1, x2, y2) face box
    is_spoof: bool = False             # this face looked like a photo/screen
    spoof_score: float = 1.0           # 1.0 = real, 0.0 = certain spoof
    appearance: str = ""               # VLM clothing/carried line (unknown only)
    expression: str = ""               # cautious mood cue (unknown only), e.g. "calm"


@dataclass
class DetectedObject:
    """A single object detected in the scene (filled in Phase 3)."""
    label: str
    confidence: float
    box: tuple = (0, 0, 0, 0)          # (x1, y1, x2, y2) in pixel coords


@dataclass
class VisitorEvent:
    # --- Identity of the event ---
    event_id: str
    timestamp: str
    trigger: str = "manual"            # "manual" | "doorbell" | "motion" | "wakeword"

    # --- Face pipeline (Phase 2 / Phase 5) ---
    identity: Identity = field(default_factory=Identity)
    face_box: tuple = (0, 0, 0, 0)
    spoof_score: float = 1.0           # 1.0 = real, 0.0 = certain spoof
    is_spoof: bool = False

    # --- Multi-person (Phase 15) ---
    # EVERY detected person, not just the largest. `identity` above still mirrors
    # the PRIMARY person (first known, else largest) so all Phase 2-14 code keeps
    # working. `extra_unknown` counts humans YOLO saw but whose face was not
    # detected (turned away / partially visible) - they are announced as "N other
    # people". Both default empty/0 so single-person events are unchanged.
    people: List["Person"] = field(default_factory=list)
    extra_unknown: int = 0

    # --- Vision pipeline (Phase 3 / Phase 6) ---
    visitor_count: int = 0
    detected_objects: List[DetectedObject] = field(default_factory=list)
    carried_objects: List[str] = field(default_factory=list)  # human-readable
    scene_summary: str = ""            # from VLM (Phase 6)
    hazards: str = "none"
    ocr_text: str = ""                 # raw OCR from parcels (Phase 6)

    # --- Visitor description (Phase 12) ---
    # age/gender are ADDITIVE fields filled from InsightFace's genderage model for
    # the largest face (known OR unknown). InsightFace age is APPROXIMATE (+/- a
    # few years), so it is only ever presented as a RANGE ("in their thirties"),
    # never as an exact number, and NEVER for known people. appearance is the VLM's
    # clothing / uniform / carried-object line, filled for UNKNOWN visitors only.
    age: Optional[int] = None          # approximate age of the largest face
    gender: str = ""                   # "man" | "woman" | "" (mapped from M/F)
    appearance: str = ""               # VLM clothing/uniform description (unknown)

    # --- Speech pipeline (Phase 7 / Phase 8) ---
    speech_transcript: str = ""
    language_detected: str = ""        # e.g. "hi", "en"
    translated_transcript: str = ""

    # --- Re-ID + memory (Phase 9) ---
    reid_id: Optional[str] = None      # stable id for repeat unknowns
    reid_seen_count: int = 0

    # --- Context engine output (Phase 3+) ---
    intent: str = "unknown visitor"
    confidence: float = 0.0
    announcement_text: str = ""        # final sentence for TTS + UI

    # --- Storage ---
    snapshot_path: str = ""

    def to_dict(self) -> dict:
        """Plain-dict view for JSON / DB. Nested dataclasses become dicts."""
        return asdict(self)


def people_to_dicts(people) -> list:
    """Serialise a list of Person (or already-dicts) to plain dicts for JSON/DB.

    Tolerant of both Person dataclasses (live pipeline) and plain dicts (rows
    rebuilt from the DB), so the same helper works on either side. Boxes stay as
    lists so json.dumps is happy.
    """
    out = []
    for p in people or []:
        if isinstance(p, Person):
            d = asdict(p)
        elif isinstance(p, dict):
            d = dict(p)
        else:                                             # pragma: no cover
            continue
        box = d.get("box")
        if isinstance(box, tuple):
            d["box"] = list(box)
        out.append(d)
    return out
